### DISCUSSION TITLE
Import only the module browser.ajax ?

### DISCUSSION THREAD

**Discussion by BobSquarePants** [brython-dev/brython#2178] [Q&A] [answered]
Hi,

I've used `brython-cli make_modules` in order to generate a .js file that only included the [browser.ajax](https://brython.info/static_doc/en/browser.ajax.html) module.

so now in my html page I've got this
```html
<script type="text/javascript" type="module" async src="brython.js"></script>
<script type="text/javascript" src="brython_ajax.js"></script>
<script type="text/python" src="test.bry"></script>   
```
and in the test.bry I got

```python
from browser import window, alert, document, ajax
```
but I have still this error in the browser console

> ModuleNotFoundError: No module named 'browser.ajax'

any ideas ?

Thanks.

**Comment by PierreQuentel** [comment, marked as answer]
I could not reproduce the issue. Here are the steps I took:
1. install Brython in an empty directory (`brython-cli install`)
2. delete **`demo.html`** (otherwise **`brython_modules.js`** has all the stdlib scripts needed by **`demo.html`**)
3. replace the content of **`index.html`** by
```xml
<!doctype html>
<html>

<head>
<meta charset="utf-8">
<script type="text/javascript" src="brython.js"></script>
<script type="text/javascript" src="brython_stdlib.js"></script>
<script type="text/python" src="test.bry"></script>
</head>

<body onload="brython(1)">
</body>

</html>
```
4. put this in **`test.bry`**
```python
from browser import window, alert, document, ajax
print('Ajax module', ajax)
```
5. for testing, run `brython-cli start_server 8001`, then load `http://localhost:8001`: it prints `Ajax module <module browser.ajax at VFS.browser.ajax.py>` in the browser console
6. close the server, then run `brython-cli make_modules`
7. in **`index.html`**, replace **`brython_stdlib.js`** by **`brython_modules.js`** (or rename it to **`brython_ajax.js`** as in your code)
8. restart the server and reload the page. I get the same message as in step 5

Can you try to follow the same sequence and tell me if you still get a ModuleNotFoundError ?

**Reply by BobSquarePants** [reply]
Hi Pierre, thx for your reply, I will test your solution this afternoon.

**Reply by BobSquarePants** [reply]
Hey @PierreQuentel ,

I followed all the steps 1-8, and it worked ! ( python 3.11.4 )

Can't wait to experiment this :) !!  

Side question I've read on https://brython.info/static_doc/en/import.html Under Optimization

>The process described above has two main drawbacks : 
>the relatively big size of brython_stdlib.js (more than 4 Mb)
>the time taken by Ajax calls 

What do you mean by `the time taken by Ajax calls ` ? is it from the `stdlib.js` ?
is with the `make_modules` to have only the ajax module  is also affected ?

**Comment by PierreQuentel** [comment]
Glad you could make it work !

Ajax calls are used for user-defined modules / packages, those that are _not_ in the standard distribution.

If an application uses both modules from the stdlib and user-defined modules:
- without optimization, you have to load **`brython_stdlib.js`** in the page for the modules in the stdlib, and the Brython execution engine sends Ajax calls for user-defined modules
- with optimization, all the modules used by the application (in stdlib + user-defined) are grouped in the single script **`brython_modules.js`** loaded in the page, so there are no more Ajax calls
